Modules/_threadmodule.c

```diff
@@ -126,49 +171,100 @@ ThreadHandle_get_ident(ThreadHandleObject *self, void *ignored)
     return PyLong_FromUnsignedLongLong(self->ident);
 }
 
+static PyObject *
+invalid_handle_error(void)
+{
+    PyErr_SetString(PyExc_ValueError,
+                    "the handle is invalid and thus cannot be detached");
+    return NULL;
+}
 
 static PyObject *
 ThreadHandle_detach(ThreadHandleObject *self, void* ignored)
 {
-    if (!self->joinable) {
-        PyErr_SetString(PyExc_ValueError,
-                        "the thread is not joinable and thus cannot be detached");
-        return NULL;
+    if (!self->is_valid) {
+        return invalid_handle_error();
     }
-    self->joinable = 0;
-    // This is typically short so no need to release the GIL
-    int ret = PyThread_detach_thread(self->handle);
-    if (ret) {
-        PyErr_SetString(ThreadError, "Failed detaching thread");
+
+    if (_PyOnceFlag_CallOnce(&self->once, (_Py_once_fn_t *)detach_thread,
+                             self) == -1) {
         return NULL;
     }
-    Py_RETURN_NONE;
+
+    switch (self->state) {
+        case THREAD_HANDLE_DETACHED: {
+            Py_RETURN_NONE;
+        }
+        case THREAD_HANDLE_JOINED: {
+            PyErr_SetString(
+                PyExc_ValueError,
+                "the thread has been joined and thus cannot be detached");
+            return NULL;
+        }
+        default: {
+            Py_UNREACHABLE();
+        }
+    }
+}
+
+static int
+join_thread(ThreadHandleObject *handle)
+{
+    int err;
+    Py_BEGIN_ALLOW_THREADS
+    err = PyThread_join_thread(handle->handle);
+    Py_END_ALLOW_THREADS
+    if (err) {
+        PyErr_SetString(ThreadError, "Failed joining thread");
+        return -1;
+    }
+    handle->state = THREAD_HANDLE_JOINED;
+    return 0;
 }
 
 static PyObject *
 ThreadHandle_join(ThreadHandleObject *self, void* ignored)
 {
-    if (!self->joinable) {
-        PyErr_SetString(PyExc_ValueError, "the thread is not joinable");
-        return NULL;
+    if (!self->is_valid) {
+        return invalid_handle_error();
     }
-    if (self->ident == PyThread_get_thread_ident_ex()) {
+
+    // We want to perform this check outside of the `_PyOnceFlag` to prevent
+    // deadlock in the scenario where another thread joins us and we then
+    // attempt to join ourselves. However, it's not safe to check thread
+    // identity once the handle's os thread has finished. We may end up with
+    // the identity stored in the handle and erroneously think we are
+    // attempting to join ourselves.
```

Comment:
For the record, it seems this is only possible on Windows. On POSIX, the pthread handle will remain valid and unique until joined or detached.

Reply:
Also, on Windows, it seems calling [GetExitCodeThread](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getexitcodethread) allows knowing if a thread identified by its handle has exited.

So perhaps we can add this following abstraction function and remove the need for `thread_is_exiting`.
```c
bool PyThread_is_current(PyThread_handle_t th);

// POSIX
bool PyThread_is_current(PyThread_handle_t th) {
  return pthread_self() == (pthread_t) *handle;
}

// Windows
bool PyThread_is_current(PyThread_handle_t th) {
  HANDLE hThread = (HANDLE) handle;
  if (GetThreadId(hThread) != GetCurrentThreadId()) {
    return false;
  }
  // The thread idents are equal, so either
  // 1. thread is the current thread
  // 2. thread was terminated and its ident reused for the current thread
  // Disambiguate by checking whether thread is still alive.
  unsigned exitcode;
  return GetExitCodeThread(hThread, &exitcode) && exitcode == STATUS_ACTIVE;
}
```

cc @gpshead

Reply:
> For the record, it seems this is only possible on Windows. On POSIX, the pthread handle will remain valid and unique until joined or detached.

This affects POSIX systems too. The situation that we're preventing against here occurs once the handle has been joined or detached. Consider the following scenario with threads T0 and T1.

1. T0 starts, finishes, and the underlying pthread is detached.
2. T1 starts and is assigned the pthread handle used by T0. This is fine because the thread underlying T0 has been detached.
3. T1 attempts to join the handle for T0. Performing the self-join check here would fail because T1 has reused the identifier for T0.

Reply:
> This affects POSIX systems too. The situation that we're preventing against here occurs once the handle has been joined or detached.

Ok, then can it simply be folded inside `join_thread`? Since you're calling `join_thread` with the once_flag, this won't be a problem anymore.

Reply:
> Ok, then can it simply be folded inside join_thread? 

No, that leaves us open to deadlocking in the following scenario:

1. T0 joins T1.
2. T1 self joins.

Reply:
Why wouldn't this prevented by the call_once flag?